### ISSUE TITLE
Is CarRacing tuned for PPO? [question]

### ISSUE THREAD

**Issue by FabianSchuetze** [DLR-RM/rl-baselines3-zoo#43]
I tried to train `CarRacing-v0` with the `ppo` algorithm with the custom hyperparameters but the final rewards seem to be far away from a good performance.  

At about `1e5` and `1e6` steps, the rollout mean reward seem to be similar, but the training losses fell during that time. I read solving the environment is defined as receiving a reward of 900 for hundred consecutive episodes. For description, I copied the two prints to stdout at 1e5 and 1e6 below.    

```Shell
-----------------------------------------
| eval/                   |             |
|    mean_ep_length       | 1e+03       |
|    mean_reward          | -82.3       |
| rollout/                |             |
|    ep_len_mean          | 962         |
|    ep_rew_mean          | -44.3       |
| time/                   |             |
|    fps                  | 101         |
|    iterations           | 25          |
|    time_elapsed         | 1004        |
|    total_timesteps      | 102400      |
| train/                  |             |
|    approx_kl            | 0.015621316 |
|    clip_fraction        | 0.0234      |
|    clip_range           | 0.4         |
|    entropy_loss         | -4.67       |
|    explained_variance   | -36.2       |
|    learning_rate        | 3e-05       |
|    loss                 | 1.81        |
|    n_updates            | 480         |
|    policy_gradient_loss | -0.00595    |
|    std                  | 0.127       |
|    value_loss           | 2.82        |
-----------------------------------------
```
and 

```Shell
-----------------------------------------
| eval/                   |             |
|    mean_ep_length       | 1e+03       |
|    mean_reward          | -82.4       |
| rollout/                |             |
|    ep_len_mean          | 1e+03       |
|    ep_rew_mean          | -75.6       |
| time/                   |             |
|    fps                  | 95          |
|    iterations           | 245         |
|    time_elapsed         | 10460       |
|    total_timesteps      | 1003520     |
| train/                  |             |
|    approx_kl            | 0.060493056 |
|    clip_fraction        | 0.148       |
|    clip_range           | 0.4         |
|    entropy_loss         | -4.06       |
|    explained_variance   | -6.8        |
|    learning_rate        | 3e-05       |
|    loss                 | 0.2         |
|    n_updates            | 4880        |
|    policy_gradient_loss | -0.015      |
|    std                  | 0.106       |
|    value_loss           | 0.465       |
-----------------------------------------
```
I trained the RL agent with the default setting by tying `python3 train.py --algo ppo --env CarRacing-v0`. 

Does anybody have an idea what I could do to train the agent better? What information could I look at? I am grateful for any tips or suggestions.

**Comment by araffin** [comment]
Hello,
I don't remember properly tuning it (and that's also why there is not trained agent yet, even in the old rl zoo), mainly because it requires a graphical interface to run :/

But I would appreciate a PR with tuned ones ;)

**Comment by FabianSchuetze** [comment]
Thanks for the kind and informative reply! 

Sounds like a nice challenge - will get back to you :-)

**Comment by FabianSchuetze** [comment]
I work on finding good hyperparameters for PPO. After 1e7 steps, the best reward is about 50~100, which is relatively far from the 900 which are considered "solving" the environment. Instead of trying to optimize the process over more steps, I was wondering if you think trying a good encoder for the pixel space through auxiliary losses is interesting? Do you think stable baselines would be interested in receiving a PR for CURL, SAC-AE, or RAD?

**Comment by araffin** [comment]
You should take a look at the hyperparameters used in https://attentionagent.github.io/
They are apparently also using frame-stacking (repo: https://github.com/xtma/pytorch_car_caring).

Paper: https://arxiv.org/abs/2003.08165

>Do you think stable baselines would be interested in receiving a PR for CURL, SAC-AE, or RAD?

This would be more for the contrib repo (not online yet).

**Comment by FabianSchuetze** [comment]
Thanks for the kind and detailed reply! I will look at all the points you mentioned and get back to you.

**Comment by FabianSchuetze** [comment]
Pre-training an AE indeed does help greatly! 

However, the results were much better when I trained a linear controller with CMA-ES as in the [world-models](https://worldmodels.github.io/) paper, then when I trained a shallow neural network with SAC.  I see that you used sac in the [5-minutes repo](https://github.com/araffin/learning-to-drive-in-5-minutes). Just out of curiosity, did you compare a linear controller trained via CMA-ES with a sac trained through backprop? 

I also tried the [pytoch_car_caring repo](https://github.com/xtma/pytorch_car_caring). They use however reward shaping to great extend which i responsible for the good score. When I use the repo without reward shaping (for 1000 epochs) the results were not impressive. The reward shaping can be seen [here](https://github.com/xtma/pytorch_car_caring/blob/1631ad17b6c7222644988d30bedd4c44bc5453ac/train.py#L57): 
```Python
        for i in range(args.action_repeat):
            img_rgb, reward, die, _ = self.env.step(action)
            # don't penalize "die state"
            if die:
                reward += 100
            # green penalty
            if np.mean(img_rgb[:, :, 1]) > 185.0:
                reward -= 0.05
            total_reward += reward
            # if no reward recently, end the episode
            done = True if self.av_r(reward) <= -0.1 else False
            if done or die:
                break
````

The self-attention paper is really fascinating, and I would love to implement it too. Do you think that would fit in stable-baslines? The approaches differ of course quiet a bit so I'm not so sure this is a good match.

**Comment by araffin** [comment]
>Pre-training an AE indeed does help greatly!

Good to hear =) (and not so [surprising](https://twitter.com/araffin2/status/1274611368499515393) ;) )

>However, the results were much better when I trained a linear controller with CMA-ES as in the world-models paper, then when I trained a shallow neural network with SAC. 

Well, did you use recommended hyperparameters for SAC? (cf doc "tips and tricks", hyperparameters matter)
And did you try stacking some frame (using the history wrapper for instance) as the environment breaks markov assumption if you give only one frame.

Yes, I did some test with CMA, it could learn to drive but then to attain maximal performance, it would take ages as it does not take every step into account.

>The self-attention paper is really fascinating, and I would love to implement it too. Do you think that would fit in stable-baselines?

We will have a contrib repo for that soon (we are currently testing it).

**Comment by FabianSchuetze** [comment]
Thanks for your kind & informative reply! 

I did use the the hyperparams suggested in the original paper, but should try different ones too. More importantly (I guess) I did not use frame stacking. CMA did not require it, so I did not use it for SAC. However, that might have helped! I will try and report again.

Great to hear a contrib repo will be public soon. I'd be happy to contribute some representation learning algorithms.
